google/cloud/storage/bucket.cc

```diff
@@ -55,6 +54,20 @@ BucketMetadata Bucket::GetMetadataImpl(
   google::cloud::internal::RaiseRuntimeError(os.str());
 }
 
+ObjectMetadata Bucket::InsertObjectMediaImpl(
+    InsertObjectMediaRequest const& request) {
+  // This operation is not idempotent, so we do not retry it.
```

Comment:
I think we could retry if we get an explicit 5XX error from the service (whereas a timeout might mean that the operation completed but the response got lost in transit).  I can't think of any scenario where the service gives a 5XX response if the state was changed; can you?

Reply:
I do not know enough about the architecture to be able to say.  In general, I am paranoid about these things because you cannot know if the state changed when you get a 500: maybe one of the middle-tiers was able to unallocate memory for the response and it sends back a 500?  Or maybe the service performs changes in multiple databases and one of the operations was successful, but the second failed?

I am not saying that GCS has any of these behaviors, is simply that I do not know.  What do other clients do?

Reply:
It looks like most of our other client libs indiscriminately retry.  I think the correct thing to do here is leave it up to the user -- if they have a reason not to retry on 5XX or timeout errors, they can use a client configured with a retry policy that doesn't retry.  However, I think in most cases, users just to be sure their upload succeeded, even if some transient edge case happened and they end up uploading it twice; to that end, I think the default client retry settings should indeed retry in the recommended error cases (https://cloud.google.com/storage/docs/json_api/v1/how-tos/simple-upload#handling_errors).